bento_beacon/utils/beacon_response.py

```diff
@@ -116,57 +60,191 @@ def received_request():
         return r
 
 
-def build_response_meta():
-    returned_schemas = g.get("response_data", {}).get("returnedSchemas", [])
-    returned_granularity = g.get("response_data", {}).get("returnedGranularity", "count")
-    received_request_summary = received_request()
+# --------------------------------------------------------------------
+#  Response control flow
+#
+#  - determine response granularity (boolean, count, record) from
+#    permissions and user request, then route response accordingly
+#
+#  - apply censorship to bool and count responses for anonymous users;
+#    full-record censorship is done elsewhere, typically by not
+#    permitting a response
+# --------------------------------------------------------------------
+
+
+def response_granularity():
+    """
+    Determine response granularity from requested granularity and permissions.
+    Returns requested granularity when requested <= max, returns max otherwise,
+    where max is the highest granularity allowed, based on this user's permissions
+    and the ordering is "boolean" < "count" < "record"
+    """
+    default_g = current_app.config["DEFAULT_GRANULARITY"].get(request.blueprint)
+    max_g = GRANULARITY_RECORD if g.permission_query_data else default_g
+    requested_g = g.request_data.get("requestedGranularity")
+
+    # below would be cleaner with an ordered enum, but that adds serialization headaches
+
+    # if max is "record" everything is permitted
+    if max_g == GRANULARITY_RECORD:
+        return requested_g
+    # if max if "boolean" nothing else is permitted
+    if max_g == GRANULARITY_BOOLEAN:
+        return max_g
```

Comment:
so if the user (with boolean access only) requests full records, they get back booleans and not an error? does the spec have an opinion on this or is it a free-for-all?

Reply:
Yeah, it's a free-for-all, although I am mostly following the suggestion [here](https://github.com/ga4gh-beacon/beacon-v2/blob/main/framework/json/requests/beaconRequestBody.json#L24) in the spec. Since we can do whatever, we could also:
- return 400 / Bad Request
- return as in this pr, but add a message in the response along the lines of "you don't have permissions for this granularity"

I think either of these is fine.

Reply:
yeah based on the spec I think this is good then!